pkg/auth/authproviders/openshift/internal/dexconnector/openshift.go

```diff
@@ -219,7 +217,20 @@ func (c *openshiftConnector) HandleCallback(s connector.Scopes, r *http.Request)
 		ctx = context.WithValue(r.Context(), oauth2.HTTPClient, c.httpClient)
 	}
 
-	token, err := c.oauth2Config.Exchange(ctx, q.Get("code"))
+	// r.URL might not get the Host part, therefore use r.Host here.
+	redirect_uri := &url.URL{
+		Scheme: "https",
```

Comment:
My biggest concern is that the way redirect URI is assembled here is different from the way it is assembled for `LoginURL()`, see e.g. https vs. http part [here](https://github.com/stackrox/stackrox/blob/d6a70fabd2b995dd3063464af8a07b71ca8b599a/pkg/auth/authproviders/openshift/backend_impl.go#L117). Because of this difference, verifying code correctness requires non-local info (openshift/backend_impl.go) and is non-trivial because possible use cases are not explicitly listed everywhere.

What seems cleaner is to _remember_ the redirect URI used in `LoginURL()` and use it in `HandleCallback()`. Maybe we can use something in `r` here to unambiguously identify the matching `LoginURL()` call. We can also extend `state` with some random string and in `LoginURL()` record the cloned config or just the redirect URI in a map keyed by `state`s. Doing so will also allow us to verify `state`, something we actually need to do anyway according to [rfc6749#section-10.12](http://tools.ietf.org/html/rfc6749#section-10.12).

Reply:
As discussed, URL reconstruction should be fine, as the callback URL shouldn't differ from the redirect_URI.
Use of `http` scheme for local endpoint has been taken into account in the appended commit.